content/ui-ux.md

```diff
@@ -0,0 +1,514 @@
+
+### Subscription readiness
+
+When you subscribe to data in Meteor, it does not become instantly available on the client. Typically the user will need to wait for at least a few hundred milliseconds (or as long as a few seconds) for the data to arrive. This is especially noticable when the app first starts up or you move between screens.
+
+There are a few UX techniques for dealing with this waiting period. The simplest is simply to switch out the page you are rendering with a generic "loading" page while you wait for all the data (typically a page may open several subscriptions) to load. As an example, in the Todos example app, we wait until all the public lists and the user's private lists have loaded before we try to render the actual page:
+
+```blaze
+{{#if Template.subscriptionsReady}}
+  {{> Template.dynamic template=main}}
+{{else}}
+  {{> appLoading}}
+{{/if}}
+```
+
+We do this with Blaze's `Template.subscriptionsReady` which is perfect for this purposes, as it waits for all the subscriptions that the current component has asked for to become ready.
+
+#### Per-component loading state
+
+Usually it makes for a better UX to show as much of the screen as possible as quickly as possible and to only show loading state for the parts of the screen that are still waiting on data. So a nice pattern to follow is "per component loading". We do this in the Todos app when you visit the list page---we instantly render the list metadata, such as it's title and privacy settings, and render a loading state for the list of todos while we wait for them to appear.
+
+[SS]
+
+We achieve this by passing the readiness of the todos list down from the smart component which is subscribing (the `listShowPage`) into the pure component which renders the data:
+
+```blaze
+{{> listsShow todosReady=Template.subscriptionsReady list=list}}
+```
+
+And then we use that state to determing what to render in the pure component (`listShow`):
+
+```blaze
+{{#if todosReady}}
+  {{#with list._id}}
+    {{#each todo in (todos this)}}
+      {{> todosItem (todoArgs todo)}}
+    {{else}}
+      <div class="wrapper-message">
+        <div class="title-message">No tasks here</div>
+        <div class="subtitle-message">Add new tasks using the field above</div>
+      </div>
+    {{/each}}
+  {{/with}}
+{{else}}
+    <div class="wrapper-message">
+      <div class="title-message">Loading tasks...</div>
+    </div>
+{{/if}}
+```
+
+#### Showing placeholders
+
+You can take the above UI a step further by showing placeholders whilst you wait for the data to load. This is a UX pattern that has been pioneered by Facebook (and that we use in Galaxy also!) which gives the user a more solid impression of what data is coming down the wire. 
+
+In the Todos app, rather than simply writing "Loading tasks..." we could place a number of placeholder tasks which give the impression of the tasks which are still yet to come down the wire.
```

Comment:
Probably should put a screenshot or a link to some article here, since we don't have it in the example app.